library/net.c

```diff
@@ -751,6 +751,9 @@ int dnet_process_recv(struct dnet_backend_io *backend, struct dnet_net_state *st
 
 void dnet_state_remove_nolock(struct dnet_net_state *st)
 {
+	if (st->__join_state == DNET_JOIN)
```

Comment:
Even more, list_move*() for state->storage_state_entry should only be performed if state->storage_state_entry is empty, and it can be replaced with list_add_tail().

This can be done because state->storage_state_entry is supposed to be added only once and into single node->storage_state_list and not moved from client to dht list like dht_state_entry is being used.